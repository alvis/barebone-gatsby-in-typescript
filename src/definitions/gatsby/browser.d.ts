/**
 *                            *** MIT LICENSE ***
 * -------------------------------------------------------------------------
 * This code may be modified and distributed under the MIT license.
 * See the LICENSE file for details.
 * -------------------------------------------------------------------------
 *
 * @summary   Definitions for gatsby-browser.ts
 *
 *            See https://www.gatsbyjs.org/docs/browser-apis
 *            for defailed usage
 *
 * @author    Alvis HT Tang <alvis@hilbert.space>
 * @license   MIT
 * @copyright Copyright (c) 2018 - All Rights Reserved.
 * -------------------------------------------------------------------------
 */

import { Action, History, Location } from 'history';
import React, { ComponentType } from 'react';

import { PageProps } from '#definitions/gatsby/page';

export interface Page {
  path: string; // any valid URL. Must start with a forward slash
  jsonName: string; // metadata of this page
  componentChunkName: string; // the chunk of this component
}

export interface PageResources {
  component: ComponentType;
  page: Page;
}

export type disableCorePrefetching = () => boolean;

export type onClientEntry = () => void;
export type onInitialClientRender = () => void;

interface OnRouteUpdate {
  action: Action;
  location: Location;
}
export type onRouteUpdate = (parameters: OnRouteUpdate) => void;
export type onPreRouteUpdate = onRouteUpdate;
export type onRouteUpdateDelayed = onRouteUpdate;

interface OnPrefetchPathname {
  pathname: string;
}
export type onPrefetchPathname = (parameters: OnPrefetchPathname) => void;

interface OnServiceWorker {
  serviceWorker: ServiceWorker;
}

export type onServiceWorkerActive = (parameters: OnServiceWorker) => void;
export type onServiceWorkerInstalled = (parameters: OnServiceWorker) => void;
export type onServiceWorkerRedundant = (parameters: OnServiceWorker) => void;
export type onServiceWorkerUpdateFound = (parameters: OnServiceWorker) => void;

export type registerServiceWorker = () => boolean;

interface ReplaceComponentRenderer<PageProps extends object> {
  props: PageProps & { pageResources: PageResources };
  loader: {
    getResourcesForPathname: (
      path: string,
      callback: (pageResources?: PageResources) => void
    ) => null | PageResources;
  };
}
export type replaceComponentRenderer<PageProps extends object = {}> = (
  parameters: ReplaceComponentRenderer<PageProps>
) => void;

// export type replaceHydrateFunction = () => (
//   element: JSX.Element,
//   container: Element,
//   callback: () => void
// ) => void;

interface ShouldUpdateScroll {
  prevRouterProps: {
    history: History;
    location: Location;
  };
  pathname: string;
}
export type shouldUpdateScroll = (parameters: ShouldUpdateScroll) => void;

interface WrapPageElement {
  element: JSX.Element;
  props: object;
}
export type wrapPageElement = (parameters: WrapPageElement) => void;

interface WrapRootElement {
  element: JSX.Element;
}
export type wrapRootElement = (parameters: WrapRootElement) => void;
