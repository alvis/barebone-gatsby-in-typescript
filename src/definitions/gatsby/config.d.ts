/**
 *                            *** MIT LICENSE ***
 * -------------------------------------------------------------------------
 * This code may be modified and distributed under the MIT license.
 * See the LICENSE file for details.
 * -------------------------------------------------------------------------
 *
 * @summary   Definitions for gatsby-config.ts
 *
 *            See https://www.gatsbyjs.org/docs/gatsby-config
 *            for defailed usage
 *
 * @author    Alvis HT Tang <alvis@hilbert.space>
 * @license   MIT
 * @copyright Copyright (c) 2018 - All Rights Reserved.
 * -------------------------------------------------------------------------
 */

import { Application as ExpressApplication } from 'express';

export type siteMetadata = object;

export interface Plugin {
  resolve: string;
  options?: object;
}

export type Plugins = Array<string | Plugin>;

export type pathPrefix = string;

export type polyfill = boolean;

export type mapping = {
  [key: string]: string;
};

export type proxy = {
  prefix: string;
  url: string;
};

export type developMiddleware = (app: ExpressApplication) => void;
