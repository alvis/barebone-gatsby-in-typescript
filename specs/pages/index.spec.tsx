/**
 *                            *** MIT LICENSE ***
 * -------------------------------------------------------------------------
 * This code may be modified and distributed under the MIT license.
 * See the LICENSE file for details.
 * -------------------------------------------------------------------------
 *
 * @summary   A test for the landing page.
 *
 * @author    Alvis HT Tang <alvis@hilbert.space>
 * @license   MIT
 * @copyright Copyright (c) 2018 - All Rights Reserved.
 * -------------------------------------------------------------------------
 */

import React from 'react';
import renderer from 'react-test-renderer';

import Index from '#pages/index';

import { IndexQuery } from '#definitions/gatsby';
import { ContextProps, PageProps, PathProps } from '#definitions/gatsby/page';

interface PagePropsGeneratorOptions<
  Query extends object | undefined = undefined
> {
  /** data from GraphQL */
  data: Query;
  /** path of the page */
  path: string;
}

/**
 * Generate dummy page props for testing pages.
 * @param options.path: the path of the page supposed to render
 */
function generatePageProps<Query extends object | undefined = undefined>(
  options?: Partial<PagePropsGeneratorOptions<Query>>
): PageProps<Query> {
  const { data, path } = {
    data: undefined,
    path: '/',
    ...options,
  };

  const contextProps: ContextProps = {
    location: { pathname: path, search: '', state: null, hash: '' },
    navigate: jest.fn(),
    pageContext: {},
    pageResources: {
      component: Index,
      page: { path: '', jsonName: '', componentChunkName: '' },
    },
    uri: path,
  };

  const pathProps: PathProps = { '*': '', path: `${path}/*` };

  return { data, ...contextProps, ...pathProps } as any;
}

describe('Index page', () => {
  const title = 'title';

  const data: IndexQuery = {
    site: {
      siteMetadata: {
        title,
      },
    },
  };

  it('should render correctly', () => {
    const indexWithData = renderer
      .create(<Index {...generatePageProps({ data })} />)
      .toJSON();
    expect(indexWithData).toMatchSnapshot();
    expect(indexWithData).toContainEqual({
      type: 'h1',
      props: {},
      children: [title],
    });

    const indexWithoutData = renderer
      .create(<Index {...generatePageProps()} />)
      .toJSON();
    expect(indexWithoutData).toMatchSnapshot();
  });
});
