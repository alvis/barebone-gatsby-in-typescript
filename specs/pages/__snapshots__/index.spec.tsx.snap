// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page should render correctly 1`] = `
Array [
  <h1>
    title
  </h1>,
  <p>
    A sitereated by Gatsby
  </p>,
]
`;

exports[`Index page should render correctly 2`] = `
<p>
  A sitereated by Gatsby
</p>
`;
